rust_snuba/ARCHITECTURE.md

```diff
@@ -0,0 +1,101 @@
+# rust-snuba consumers
+
+Snuba has two ways to ingest data into clickhouse: The old, Python-based consumers, and this Rust consumer.
+
+This document is about the Rust one, but due to it starting out as a
+line-by-line port of the Python consumer, the higher-level architecture also
+applies to Python.
+
+## Entrypoint
+
+`rust_snuba` is a library crate that is being compiled with PyO3/maturin into a
+Python wheel. This Python wheel exposes a `consumer` function which is used
+from within the Snuba (Python) CLI to launch `snuba rust-consumer`.
+Configuration parsing, CLI parsing and runtime config are still handled in
+Python so as to avoid unnecessary code duplication and potentially multiple
+sources of truth for config.
+
+## Steps
+
+`rust_snuba` then uses the Rust port of Arroyo to define its own consumer. The following steps (processing strategies) exist (see `factory.rs`):
+
+1. `HealthCheck` step, touches a file in irregular intervals to signal to k8s
+   that the consumer is still alive.
+2. `SetJoinTimeout` step, resets the join timeout to 0 for the next steps. This
+   marks a section where it's ok to drop in-flight progress during rebalancing.
```

Comment:
Am I understanding correctly:

Message validation and transformation and opening a batch (i.e. opening a connection with CH) uses a join timeout = 0 as this in-flight work can just be "dropped". Or in the case of the Reduce/opening a batch, the connection is still open, so the join timeout has no "impact" in that sense. How do we guarantee idempotency though? i.e. what if the consumer restarts, and rewrites some of the same rows to the socket?

Reply:
you can write to the socket, and then terminate the connection, and clickhouse will not actually write the rows. in that sense writing to the socket is idempotent. only when you explicitly close the batch as per HTTP chunked transfer, the batch gets written by clickhouse.

Reply:
note: add clarification to document